src/eos-updater-poll.c

```diff
@@ -162,105 +135,253 @@ read_config (const gchar *config_file_path,
+  g_autoptr(GPtrArray) finders = NULL;  /* (element-type OstreeRepoFinder) */
+  g_autoptr(RepoFinderAvahiRunning) finder_avahi = NULL;
+  gboolean redirect_followed = FALSE;
+
+  if (!get_booted_refspec (&booted_refspec, NULL, NULL, &booted_collection_ref, error))
+    return NULL;
+
+  if (booted_collection_ref == NULL)
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+                   "No collection ID set for currently booted deployment.");
+      return NULL;
+    }
+
+  upgrade_collection_ref = booted_collection_ref;
+  upgrade_refspec = booted_refspec;
+
+  finders = get_finders (config, context, &finder_avahi);
+
+  /* Check whether the commit is a redirection; if so, fetch the new ref and
+   * check again. */
+  do
+    {
+      const OstreeCollectionRef *refs[] = { NULL, NULL };
+      g_autoptr(GVariant) pull_options = NULL;
+      g_autoptr(GAsyncResult) find_result = NULL, pull_result = NULL;
+      g_auto(GVariantBuilder) builder = G_VARIANT_BUILDER_INIT (G_VARIANT_TYPE ("a{sv}"));
+
+      refs[0] = upgrade_collection_ref;
+      g_debug ("%s: Finding remotes advertising upgrade_collection_ref: (%s, %s)",
+               upgrade_collection_ref->collection_id, upgrade_collection_ref->ref_name);
+
+      ostree_repo_find_remotes_async (repo, refs, NULL  /* options */,
+                                      (OstreeRepoFinder **) finders->pdata,
+                                      NULL  /* progress */,
+                                      cancellable, async_result_cb, &find_result);
+
+      while (find_result == NULL)
+        g_main_context_iteration (context, TRUE);
+
+      results = ostree_repo_find_remotes_finish (repo, find_result, error);
+      if (results == NULL)
+        return NULL;
+
+      /* No updates available. */
+      if (results[0] == NULL)
+        {
+          g_message ("Poll: Couldn’t find any updates");
+          return NULL;
+        }
+
+      g_variant_builder_add (&builder, "{s@v}", "flags",
+                             g_variant_new_variant (g_variant_new_int32 (OSTREE_REPO_PULL_FLAGS_COMMIT_ONLY)));
+      pull_options = g_variant_ref_sink (g_variant_builder_end (&builder));
+
+      ostree_repo_pull_from_remotes_async (repo,
+                                           (const OstreeRepoFinderResult * const *) results,
+                                           pull_options, NULL  /* progress */, cancellable,
+                                           async_result_cb, &pull_result);
+
+      while (pull_result == NULL)
+        g_main_context_iteration (context, TRUE);
+
+      if (!ostree_repo_pull_from_remotes_finish (repo, pull_result, error))
+        return NULL;
+
+      g_clear_pointer (&new_refspec, g_free);
+      g_clear_pointer (&new_collection_ref, ostree_collection_ref_free);
+
+      /* Parse the commit and check there’s no redirection to a new ref. */
+      if (!parse_latest_commit (repo, upgrade_refspec, &redirect_followed,
```

Comment:
Similar to the other case, I think we're leaking `checksum`

Reply:
Yup, fixup will be pushed shortly.